File: sigpy_recon/test_llr.py
import torch
import cupy as cp
import sigpy as sp
import numpy as np
import nibabel as nib


def _llr(x, lamda, shift):
    """
    x: [Nt, Nx, Ny, Nz]
    """
    shape = x.shape
    block = 5
    stride = block
    N = len(x.shape[1:])
    B = sp.linop.ArrayToBlocks(shape, (block,)*N, (stride,)*N)
    print(B.oshape)
    T = sp.linop.Transpose(B.oshape, (1, 2, 3, 0, 4, 5, 6))
    n = T.oshape[0] * T.oshape[1] * T.oshape[2]
    R = sp.linop.Reshape((n, shape[0], block**N), T.oshape)
    L = R * T * B
    device = sp.get_device(x)
    xp = device.xp
    with device:
        for k in range(N):
            x = xp.roll(x, shift[k], axis=-(k + 1))
    mats = L(x)
    (u, s, vh) = xp.linalg.svd(mats, full_matrices=False)
    thresh_s = s - lamda
    thresh_s[thresh_s < 0] = 0
    print(u.shape, thresh_s.shape, vh.shape)
    tmp = u * thresh_s[..., None, :]
    tmp2 = xp.sum(tmp[...,None] * vh[:,None,:,:], axis=-2)
    print(mats.shape, tmp2.shape)
    mats[...] = tmp2
    x = L.H(mats)
    for k in range(N):
        x = xp.roll(x, -shift[k], axis=-(k + 1))
    return x
  
if __name__ == '__main__':
    niiimg = nib.load('/home/fs0/qijia/scratch/moco_exp/expout/subspace_25-10-23/rd_61.nii.gz')
    img = niiimg.get_fdata() * 1e10
    print(img.shape)
    img = sp.to_device(img, 0).transpose(3, 0, 1, 2)
    x = _llr(img, 100, (0, 0, 0))
    x = sp.to_device(x, -1).transpose(1, 2, 3, 0)
    nib.save(nib.Nifti1Image(x, niiimg.affine), '/home/fs0/qijia/scratch/moco_exp/expout/subspace_25-10-23/rd_61_llr_sigpy.nii.gz')

File: sigpy_recon/test_llr.m
[img, ~, scales, ~, ~] = read_avw('/home/fs0/qijia/scratch/moco_exp/expout/subspace_25-10-23/rd_61.nii.gz');
img = img*1e10;
im_size = size(img);
s1 = 25; s2 = 30; s3 = 0;
p=[5,5,5];
lambda = 100;
b = 1;
[ii,jj,kk]  =   meshgrid(s1-(p(1)-1)/2:p(1):im_size(1), s2-(p(2)-1)/2:p(2):im_size(2), s3-(p(3)-1)/2:p(3):im_size(3));

x = zeros(size(img));

for idx = 1:length(ii(:))
    q   =   get_patch(img, ii(idx), jj(idx), kk(idx), p);
    [u,s,v]     =   svd(reshape(q,[],im_size(4)),'econ');
    s   =   shrink(s, lambda*b); 
    q   =   reshape(u*s*v', size(q));
    x   =   put_patch(x, q, ii(idx), jj(idx), kk(idx), p);
end

save_avw(x, '/home/fs0/qijia/scratch/moco_exp/expout/subspace_25-10-23/rd_61_llr_matlab.nii.gz', 'd', scales);


function q = get_patch(X, i, j, k, p)

    [sx,sy,sz,st]   =   size(X);
    q               =   X(max(i-(p(1)-1)/2,1):min(i+(p(1)-1)/2,sx),max(j-(p(2)-1)/2,1):min(j+(p(2)-1)/2,sy), max(k-(p(3)-1)/2,1):min(k+(p(3)-1)/2,sz),:);
    
end

function X = put_patch(X, q, i, j, k, p)
    [sx,sy,sz,st]   =   size(X);
    X(max(i-(p(1)-1)/2,1):min(i+(p(1)-1)/2,sx),max(j-(p(2)-1)/2,1):min(j+(p(2)-1)/2,sy), max(k-(p(3)-1)/2,1):min(k+(p(3)-1)/2,sz),:) = q;
end

function y = shrink(x, thresh)
    y = diag(max(diag(x)-thresh,0));
end